src/volatile_memory.rs

```diff
@@ -671,6 +679,223 @@ impl<'a, T: ByteValued> VolatileRef<'a, T> {
+        self.nelem == 0
+    }
+
+    /// Gets the size of this slice.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use vm_memory::{VolatileRef, VolatileArrayRef};
+    ///   let v_array = unsafe { VolatileArrayRef::<u32>::new(0 as *mut _, 1) };
+    ///   assert_eq!(v_array.len(), 1);
+    /// ```
+    pub fn len(&self) -> usize {
+        self.nelem
+    }
+
+    /// Gets the element size of this slice.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use std::mem::size_of;
+    /// # use vm_memory::VolatileRef;
+    ///   let v_ref = unsafe { VolatileRef::<u32>::new(0 as *mut _) };
+    ///   assert_eq!(v_ref.len(), size_of::<u32>() as usize);
+    /// ```
+    pub fn element_size(&self) -> usize {
+        size_of::<T>()
+    }
+
+    /// Gets the address of this slice's memory.
+    pub fn as_ptr(&self) -> *mut u8 {
+        self.addr
+    }
+
+    /// Converts this `T` reference to a raw slice with the same size and address.
+    pub fn to_slice(&self) -> VolatileSlice<'a> {
+        unsafe { VolatileSlice::new(self.addr, self.nelem * self.element_size()) }
+    }
+
+    /// Does a volatile read of the I-th value pointed to by this ref.
+    pub fn ref_at(&self, index: usize) -> VolatileRef<'a, T> {
+        // Safe because the memory has the same lifetime and points to a subset of the
+        // memory of the VolatileArrayRef.
+        unsafe {
+            let byteofs = (self.element_size() * index) as isize;
+            let ptr = self.as_ptr().offset(byteofs);
+            VolatileRef::new(ptr)
+        }
+    }
+
+    /// Does a volatile read of the I-th value pointed to by this ref.
+    pub fn load(&self, index: usize) -> T {
+        self.ref_at(index).load()
+    }
+
+    /// Does a volatile write of the I-th value pointed to by this ref.
+    pub fn store(&self, index: usize, value: T) {
+        self.ref_at(index).store(value)
+    }
+
+    /// Copies `self.len()` or `buf.len()` times the size of `T` bytes, whichever is smaller, to
+    /// `buf`.
+    ///
+    /// The copy happens from smallest to largest address in `T` sized chunks using volatile reads.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use std::fs::File;
+    /// # use std::path::Path;
+    /// # use vm_memory::VolatileMemory;
+    /// # fn test_write_null() -> Result<(), ()> {
+    /// let mut mem = [0u8; 32];
+    /// let mem_ref = &mut mem[..];
+    /// let vslice = mem_ref.get_slice(0, 32).map_err(|_| ())?;
+    /// let mut buf = [5u8; 16];
+    /// vslice.copy_to(&mut buf[..]);
+    /// for v in &buf[..] {
+    ///     assert_eq!(buf[0], 0);
+    /// }
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn copy_to(&self, buf: &mut [T]) -> usize {
+        let mut addr = self.addr;
+        let mut i = 0;
+        for v in buf.iter_mut().take(self.len()) {
+            unsafe {
+                *v = read_volatile(addr as *const Packed<T>).0;
+                addr = addr.add(self.element_size());
```

Comment:
Seems like `ptr::add` also states undefined behavior might happen if the result overflows an `isize`.

Reply:
Here there is an implicit check against self.len() (thanks to .take(self.len())) so it's enough to check that the size fits an isize in get_array_ref.